Assets/MRTK/Core/Services/BaseService.cs

```diff
@@ -68,12 +73,19 @@ public virtual void Destroy()
 
         private bool? isInitialized = null;
 
+        private readonly string typeName = null;
+
         /// <inheritdoc />
         public virtual bool IsInitialized
         {
             get
             {
-                Debug.Assert(isInitialized.HasValue, $"{GetType()} has not set a value for IsInitialized, returning false.");
+                if (!isInitialized.HasValue)
+                {
+                    // Calling this allocates a string, so test the condition before
+                    Debug.AssertFormat(isInitialized.HasValue, "{0} has not set a value for IsInitialized, returning false.", typeName);
```

Comment:
in this case, why use Assert? that just causes the condition to be evaluated twice?

Reply:
The issue with these has been that the interpolated strings alloc even if the assert isn't hit (so, adding the `if` means we avoid the assert unless we know it fails, which does feel roundabout!). I left a separate comment (missed yours before I posted!) wondering if the move to AssertFormat will do enough to avoid the alloc, making the `if` unnecessary.

Reply:
https://github.com/microsoft/MixedRealityToolkit-Unity/pull/10310#discussion_r747153589

Reply:
Yeah, honestly what I would prefer here is to gate the assert in it's own function MRTKDiagAssert, and have it be conditionally compiled based upon some #define that can be turned on.  As this is a really low level property that gets called multiple times per frame per service.  And the number of times that that assert has helped me as an external MRTK user is zero.